pennylane/beta/tapes/qnode.py

```diff
@@ -0,0 +1,541 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the QNode class and qnode decorator.
+"""
+from collections.abc import Sequence
+from functools import lru_cache, update_wrapper
+import warnings
+
+import numpy as np
+
+from pennylane import Device
+from pennylane.beta.queuing import MeasurementProcess
+from pennylane.beta.tapes import QuantumTape
+from pennylane.beta.interfaces.autograd import AutogradInterface
+
+
+class QuantumFunctionError(Exception):
+    """Exception raised when an illegal operation is defined in a quantum function."""
+
+
+class QNode:
+    """Base class for quantum nodes in the hybrid computational graph.
+
+    A *quantum node* encapsulates a :ref:`quantum function <intro_vcirc_qfunc>`
+    (corresponding to a :ref:`variational circuit <glossary_variational_circuit>`)
+    and the computational device it is executed on.
+
+    The QNode calls the quantum function to construct a :class:`~.QuantumTape` instance representing
+    the quantum circuit.
+
+    .. note::
+
+        As the quantum tape is a *beta* feature, the standard PennyLane
+        measurement functions cannot be used. You will need to instead
+        import modified measurement functions within the quantum tape:
+
+        >>> from pennylane.beta.queuing import expval, var, sample, probs
+
+    Args:
+        func (callable): a quantum function
+        device (~.Device): a PennyLane-compatible device
+        interface (str): The interface that will be used for classical backpropagation.
+            This affects the types of objects that can be passed to/returned from the QNode:
+
+            * ``interface='autograd'``: Allows autograd to backpropogate
+              through the QNode. The QNode accepts default Python types
+              (floats, ints, lists) as well as NumPy array arguments,
+              and returns NumPy arrays.
+
+            * ``interface='torch'``: Allows PyTorch to backpropogate
+              through the QNode. The QNode accepts and returns Torch tensors.
+
+            * ``interface='tf'``: Allows TensorFlow in eager mode to backpropogate
+              through the QNode. The QNode accepts and returns
+              TensorFlow ``tf.Variable`` and ``tf.tensor`` objects.
+
+            * ``None``: The QNode accepts default Python types
+              (floats, ints, lists) as well as NumPy array arguments,
+              and returns NumPy arrays. It does not connect to any
+              machine learning library automatically for backpropagation.
+
+        diff_method (str, None): the method of differentiation to use in the created QNode.
+
+            * ``"best"``: Best available method. Uses classical backpropagation or the
+              device directly to compute the gradient if supported, otherwise will use
+              the analytic parameter-shift rule where possible with finite-difference as a fallback.

```

Comment:
Is `"reversible"` not attempted?

Reply:
Not currently. This reflects core. If we add it to `'best'`, we should do it in a separate PR.

@co9olguy, should we add it to the `'best'` hierarchy? or is it too dependent on circuit structure, so best to leave it to the user to explicitly request?